pkg/controller/volume/persistentvolume/scheduler_binder.go

```diff
@@ -382,13 +436,33 @@ func (b *volumeBinder) checkBindings(pod *v1.Pod, bindings []*bindingInfo, claim
 		return false, fmt.Errorf("failed to get cached claims to provision for pod %q", podName)
 	}
 
+	// Check for any conditions that might require scheduling retry
+
 	for _, binding := range bindings {
-		// Check for any conditions that might require scheduling retry
+		// Check against latest API object
```

Comment:
Can you explain why we need to check the API object directly and not wait for the informer update to the pv cache?

Reply:
Because PV cache may be overridden. Binding operation is async, scheduling in main loop may Find/Assume new cache on new API objects, then binding operation will don't know latest API objects. This is root cause of volume binder races (e.g. https://github.com/kubernetes/kubernetes/issues/71928). 

e.g. 

- Volume binder find/assume PV/PVC cache (cache: prebound, api: not bound)
- Volume binder make API call to bind PV/PVC and check PV & PVC is bound or not (cache: prebound, api: prebound)
- PV/PVC is rejected (cache: not bound, api: not bound)
- Before binding operation check PV/PVC cache, Volume binder find/assume PV/PVC cache again (cache: prebound, api: not bound)
- Binding operation finds PV/PVC cache are still the same, but actually binding operation should be restarted to bind PV/PVC again

In current scheduler, we can simply do not share PV/PVC cache in two goroutines to avoid races. IMO, PV/PVC cache are used for scheduling, not for binding operation check which should depends on latest state in API server. Because PV/PVC are bound or not and pod should be rescheduled or not depend on state in API server which are updated by PV controller or provisioner. 

Though PV/PVC cache can be used if they will not be modified by another one if we prevent scheduler from scheduling pod when pod is in binding , but it is still unnecessary.

Reply:
Ok that makes sense. Can you update the comment here with a little more detail, saying that because the PV/PVC can be assumed again, we need to fail the binding operation so it can be retried.

And actually, assume pod will fail because binding is in progress.  After this binding operation fails, then what causes the pod to retry scheduling again?

Reply:
Yes, if pod is not marked unschedulable (status=PodReasonUnschedulable) wrongly (fixed in this PR), [other status updates will re-add pod into active queue](https://github.com/kubernetes/kubernetes/blob/5b6e20d0e057c0060a185673adc75af44d1a4566/pkg/scheduler/internal/queue/scheduling_queue.go#L367). This will retry scheduling again.

Reply:
assume pod [fails with non-PodReasonUnschedulable error](https://github.com/kubernetes/kubernetes/blob/5b6e20d0e057c0060a185673adc75af44d1a4566/pkg/scheduler/scheduler.go#L410-L416) which will not put pod into unscheduable pod.

Reply:
Comments are updated in 80d8815f64fb1452d8c62c2513707b90401d2b10.